src/GraphQL/Types/Composite/InputObjectGraphType.cs

```diff
@@ -44,5 +58,23 @@ public virtual object ParseDictionary(IDictionary<string, object> value)
             // for InputObjectGraphType<TSourceType>, convert to TSourceType via ToObject.
             return value.ToObject(typeof(TSourceType), this);
         }
+
+        /// <inheritdoc/>
+        public virtual bool IsValidDefault(object value) => value is TSourceType;
+
+        /// <summary>
+        /// Converts a value to an AST representation. This is necessary for introspection queries
+        /// to return the default value for fields of this scalar type. This method may throw an exception
+        /// or return null for a failed conversion.
+        /// <br/><br/>
+        /// The default implementation always throws an exception. It is recommended that this method be
+        /// overridden to support introspection of fields of this type that have default values. This method
+        /// is not otherwise needed to be implemented.
+        /// </summary>
+        public virtual IValue ToAst(object value)
+        {
+            //TODO: use reflection to reverse-engineer TSourceType and provide a default implementation
+            throw new System.NotImplementedException($"Please override the '{nameof(ToAst)}' method of '{Name}' to support this operation.");
+        }
```

Comment:
Why so? What about delated code?
```csharp
 if (!(value is Dictionary<string, object> dict))
                {
                    return null;
                }

                var fields = dict
                    .Select(pair =>
                    {
                        var fieldType = input.GetField(pair.Key)?.ResolvedType;
                        return new ObjectField(pair.Key, AstFromValue(pair.Value, schema, fieldType));
                    })
                    .ToList();

                return new ObjectValue(fields);
```

My only suspicion why you did this is that the `value` was never actually `Dictionary<string, object>`. Was it ?

Reply:
I doubt it would ever be so.  For code-first, you'd typically write `field.DefaultValue = new MyClass { Id = 5 };`, not `field.DefaultValue = new Dictionary<string, object> { { "Id", 5 } };`.

Now maybe for schema-first.....  but I don't know about that.

What do you think?  Should I restore the old code in there?

Reply:
```csharp
protected virtual QueryArgument ToArgument(GraphQLInputValueDefinition inputDef)
        {
            return new QueryArgument(ToGraphType(inputDef.Type))
            {
                Name = (string)inputDef.Name.Value,
                DefaultValue = inputDef.DefaultValue.ToValue(),   <---
                ResolvedType = ToGraphType(inputDef.Type),
                Description = inputDef.Comment?.Text.ToString()
            }.SetAstType(inputDef);
        }
```

Then look at ` public static object ToValue(this GraphQLValue source)`:
```csharp
  case ASTNodeKind.ObjectValue:
                {
                    var obj = source as GraphQLObjectValue;
                    var values = new Dictionary<string, object>();  <--- VIOLA!

                    Debug.Assert(obj != null, nameof(obj) + " != null");
                    if (obj.Fields != null)
                    {
                        foreach (var f in obj.Fields)
                            values[(string)f.Name.Value] = ToValue(f.Value);
                    }

                    return values;
                }
```

So I suggest to return this code but a little changed:
```csharp
if (!(value is Dictionary<string, object> dict))
{
    return inputObject.ToAst(value);         <----- CHANGED
}

var fields = dict
    .Select(pair =>
    {
        var fieldType = input.GetField(pair.Key)?.ResolvedType;
        return new ObjectField(pair.Key, AstFromValue(pair.Value, schema, fieldType));
    })
    .ToList();

return new ObjectValue(fields);
```

Of course change it appropriate, I just copy-pasted it from my previous comment.

Reply:
What does inputObject.ToAST to? Call itself? This IS the inputObject.ToAST method we are discussing.

Reply:
I think what you’re trying to say is that this would still be useful for schema-first implementations, even if not code-first. But adding inputObject.ToAST isn’t applicable.

Reply:
> What does inputObject.ToAST to? Call itself? This IS the inputObject.ToAST method we are discussing.

I wrote about deleted code from the extension method - `  public static IValue GraphQLExtensions.ToAST(this IGraphType type, object value)`
![изображение](https://user-images.githubusercontent.com/21261007/109396226-a13e9380-7941-11eb-853a-9d849ded3971.png)

> I think what you’re trying to say is that this would still be useful for schema-first implementations, even if not code-first.

Both for schema-first (when Dictionary) and graphtype-first (when custom object).

> But adding inputObject.ToAST isn’t applicable.

Please check again.

Reply:
I’ll check it when I’m at a PC. I don’t know what you’re referring to. The extension method we created just calls this method for input objects. If I recall correctly, this would just create an infinite loop.

Reply:
I talked about returning "old" code into extension method, not into this virtual method. Virtual method should be as is now - generating `NotImplementedException`. So user **must** implement it when working with custom default values in graphtype-first like `argument.default = new new CustomObject { Name = "Tom" }`

Reply:
Hmm. Ok so then we are providing a default implementation for schema first users, where the value is coerced to a dictionary. Shame that there’s a double conversion. Similar issue for custom scalars or any other object - for schema first we don’t need to double do conversion if we store the original AST. Perhaps we can restore DefaultValueAST for this need. Then we can eliminate this code. ...????

Reply:
> Ok so then we are providing a default implementation for schema first users, where the value is coerced to a dictionary

Yes.

> Shame that there’s a double conversion.

Where? Please explain.

Reply:
Schema builder takes an AST with the default value, converts it to a dictionary and stores it in DefaultValue, which is used by ToAST to conversion back to an AST for introspection.

Reply:
Not a big deal.

Reply:
Of course we **can** cache AST in internal `FieldType` property and reset it in `DefaultValue.set`. The same for QueryArgument.

Reply:
#2336

Reply:
ok